Fundamentals/Setup/Server-Setup/runtime-modes.md

```diff
@@ -0,0 +1,83 @@
+---
+versionFrom: 10.1.0
+---
+
+# Runtime modes
+
+_This section describes best practices with running Umbraco in production._
+
+## What are runtime modes?
+
+Runtime modes are settings you can enable via your Appsettings.json, which enforce best practices for different styles of development or production.
+
+There are 3 available runtime modes in Umbraco:
+
+- BackofficeDevelopment
+- Development
+- Production
+
+## BackofficeDevelopment mode
+
+BackofficeDevelopment mode is the default behavior for Umbraco and it does not enforce any settings at all.
+
+# Development mode
+
+Development mode is made for code development, this mode makes you change your ModelsBuilder mode to "SourceCodeAuto" or "SourceCodeManual".
+In this mode your Razor views will still automatically be recompiled, but if you change your document types you need to manually recompile your application.
+
+To enable development mode, first change your `appsettings.Development.json` to include the following settings:
+
+```json
+{
+    "Umbraco": {
+        "CMS": {
+            "Runtime": {
+                "Mode" : "Development"
+            },
+            "ModelsBuilder":{
+                "ModelsMode": "SourceCodeAuto"
+            }
+        }
+    }
+}
+```
+
+Now ensure your models are generated by running Umbraco and going to Settings > Models Builder > Generate models.
+You're now able to remove the following from your csproj file:
+
+```c#
+    <RazorCompileOnBuild>false</RazorCompileOnBuild>
+    <RazorCompileOnPublish>false</RazorCompileOnPublish>
+    <CopyRazorGenerateFilesToPublishDirectory>true</CopyRazorGenerateFilesToPublishDirectory>
+```
+
+If you run the application, it should now be in development mode.
+
+# Production mode
+
+Production mode is made for running your site in production, this mode enforces you to have set "UseHttps" to true, and "UmbracoApplicationUrl" to the canonical url of your Umbraco installation.
+
+To enable production mode, add a `appsettings.Production.json` file to your project with the following settings: 
+
+```json
+{
+  "Umbraco": {
+    "CMS": {
+      "Runtime": {
+        "Mode": "Production"
+      },
+      "Global": {
+        "UseHttps": true
+      },
+      "ModelsBuilder": {
+        "ModelsMode": "Nothing"
+      },
+      "WebRouting": {
+        "UmbracoApplicationUrl": "https://localhost:5001/"
+      }
+    }
+  }
+}
+```
+
+Do note to use "ModelsMode": "Nothing" you have to have already generated your models before setting this.
```

Comment:
```suggestion

Although you can still edit document types and views (if not running from the published output), changes won't be picked up until you've rebuilt your project or republished the application. Also note that models won't be generated by ModelsBuilder (because the mode is set to `Nothing`), requiring your to do all your changes while in `Development` mode.

Also ensure the `UmbracoApplicationUrl` is updated to the primary URL of your production environment, as this is used when sending emails (password reset, notifications, health check results, etc.) and the keep-alive task.

## Customize/extend runtime mode validation

Validation of the above mentioned settings are done when determining the runtime level during startup using the new `IRuntimeModeValidationService` and when it fails, causes a `BootFailedException` to be thrown. The default implementation gets all registered `IRuntimeModeValidators` to do the validation, making it possible to remove default checks and/or add your own (inherit from `RuntimeModeProductionValidatorBase` if you only want to validate against the production runtime mode). The following validators are added by default:
- `JITOptimizerValidator` - Ensure the application is built/published in Release mode (with JIT optimization enabled) when in production runtime mode, e.g. using `dotnet publish --configuration Release`;
- `UmbracoApplicationUrlValidator` - ensure `Umbraco:CMS:WebRouting:UmbracoApplicationUrl` is configured when in production runtime mode;
- `UseHttpsValidator` - ensure `Umbraco:CMS:Global:UseHttps` is enabled when in production runtime mode;
- `RuntimeMinificationValidator` - ensure `Umbraco:CMS:RuntimeMinification:CacheBuster` is set to a fixed cache buster like `Version` or `AppDomain` when in production runtime mode;
- `ModelsBuilderModeValidator` - ensure `Umbraco:CMS:ModelsBuilder:ModelsMode` is not set to `InMemoryAuto` when in development runtime mode and set to `Nothing` when in production runtime mode.
```

The last part can be updated when PR https://github.com/umbraco/Umbraco-CMS/pull/12693 is merged, making it a much nicer experience to remove default runtime mode validators.